### PR TITLE
Final merge to main for Key Vault 7.3

### PR DESCRIPTION
**Pull request by gearama**
Bringing the code from the feature branch into main for KV 7.3

# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

File: eng/pipelines/templates/stages/platform-matrix-live.json [hunk 1]
```diff
@@ -128,12 +128,14 @@
                     "RunSamples": 1,
                     "WindowsCtestConfig": "-C Release"
                 },
-                "Win_x64_with_unit_test_libcurl": {
-                    "VCPKG_DEFAULT_TRIPLET": "x64-windows-static",
-                    "CMAKE_GENERATOR_PLATFORM": "x64",
-                    "CmakeArgs": " -DBUILD_TRANSPORT_CURL=ON -DBUILD_TESTING=ON -DRUN_LONG_UNIT_TESTS=ON -DBUILD_PERFORMANCE_TESTS=ON -DMSVC_USE_STATIC_CRT=ON ",
-                    "BuildArgs": "-v --parallel 8 --config Release",
-                    "WindowsCtestConfig": "-C Release"
+            "HSM_Win_x64_with_unit_test_libcurl": {
```
Comment (ahsonkhan) [inline eng/pipelines/templates/stages/platform-matrix-live.json:131]: nit: spacing/indentation

File: sdk/keyvault/azure-security-keyvault-certificates/CHANGELOG.md [hunk 2]
```diff
@@ -4,11 +4,13 @@
...
-### Breaking Changes
+- Keyvault 7.3 support added for Certificates.
 
-### Bugs Fixed
+## 4.0.0 (2022-06-07)
 
-### Other Changes
+### Breaking Changes
+
+- Renamed `keyvault_certificates.hpp` to `certificates.hpp`.
```
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-certificates/CHANGELOG.md:14]: Looks like a bad merge, 4.0.0. is duplicated here.

File: sdk/keyvault/azure-security-keyvault-certificates/inc/azure/keyvault/certificates/certificate_client_options.hpp [hunk 3]
```diff
@@ -18,59 +18,23 @@
...
    *
    */
   struct CertificateClientOptions final : public Azure::Core::_internal::ClientOptions
   {
-    ServiceVersion Version;
+    /**
+     * @brief Service Version used.
+     *
+     */
+    std::string Version;
```
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-certificates/inc/azure/keyvault/certificates/certificate_client_options.hpp:31]: All the changes around removing `ServiceVersion` and leaving things as `string` seem to be breaking changes. Update the changelog to call that out.

Also, why did we decide to do it this way, rather than using the `ServiceVersion` class?

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client.hpp [hunk 4]
```diff
@@ -398,11 +419,70 @@ namespace Azure { namespace Security { namespace KeyVault { namespace Keys {
...
+        Azure::Core::Context const& context = Azure::Core::Context()) const;
+
+    /**
+     * @brief Get the requested number of bytes containing random values.
+     *
+     * @remark Get the requested number of bytes containing random values from a managed HSM.
+     *
+     * @param options The request object to get random bytes.
+     * @param context A #Azure::Core::Context controlling the request lifetime.
+     * @return Azure::Response<std::vector<uint8_t>>
```
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client.hpp:480]: Comment and API mismatch

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp [hunk 5]
```diff
@@ -120,8 +121,8 @@ namespace Azure { namespace Security { namespace KeyVault { namespace Keys {
     AZ_SECURITY_KEYVAULT_KEYS_DLLEXPORT static const KeyOperation WrapKey;
 
     /**
-     * @brief The key can be used to unwrap another key with the UnwrapKey(KeyWrapAlgorithm, Byte[],
-     * CancellationToken) method.
+     * @brief The key can be used to unwrap another key with the UnwrapKey(KeyWrapAlgorithm,
+     * Byte[], CancellationToken) method.
```
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp:125]: Should the documentation really say `CancellationToken` here? We should update this comment to refer to the correct C++ method:
https://github.com/Azure/azure-sdk-for-cpp/blob/8a12a646290505a41bc8fe0134b4c5558b209e00/sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/cryptography/cryptography_client.hpp#L156-L159

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp [hunk 6]
```diff
@@ -130,6 +131,12 @@ namespace Azure { namespace Security { namespace KeyVault { namespace Keys {
      * CancellationToken) method.
      */
     AZ_SECURITY_KEYVAULT_KEYS_DLLEXPORT static const KeyOperation Import;
+
+    /**
+     * @brief The key can be exported during creation using the ImportKey(ImportKeyOptions,
+     * CancellationToken) method.
```
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp:137]: Same issue here and elsewhere in this file:
https://github.com/Azure/azure-sdk-for-cpp/blob/2feed76d860eaa5ac4697f6a235337a54285154d/sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp#L92-L132

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp [hunk 7]
```diff
@@ -292,6 +299,39 @@ namespace Azure { namespace Security { namespace KeyVault { namespace Keys {
     AZ_SECURITY_KEYVAULT_KEYS_DLLEXPORT static const KeyCurveName P521;
   };
 
+  /**
+   * @brief The encryption algorithm to use to protected the exported key material.
```
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp:303]: This comment is coming from the swagger, but we should fix it to be worded correctly:
https://github.com/Azure/azure-rest-api-specs/blob/37cd8dfac3c570a24bb645b31c012d12efb760df/specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/keys.json#L1772

```suggestion
   * @brief The encryption algorithm to use to protect the exported key material.
```

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_options.hpp [hunk 8]
```diff
@@ -43,55 +43,23 @@ namespace Azure { namespace Security { namespace KeyVault { namespace Keys {
     Azure::Nullable<std::string> NextPageToken;
   };
 
-  class ServiceVersion final {
```
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_options.hpp:46]: This will be considered a breaking change, so we should really explore why it's worth doing.

File: sdk/keyvault/azure-security-keyvault-keys/src/key_release_options.cpp [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+{
+  ReleaseKeyResult result;
+
+  auto const& body = rawResponse.GetBody();
+  auto jsonParser = Azure::Core::Json::_internal::json::parse(body);
+
+  result.Value = jsonParser[_detail::ValueParameterValue].get<std::string>();
+
+  return result;
+}
```
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-keys/src/key_release_options.cpp:50]: nit: add a new line at eof

File: sdk/keyvault/azure-security-keyvault-keys/test/samples/sample1-hello-world/CMakeLists.txt [hunk 10]
```diff
@@ -14,4 +14,4 @@ add_executable (
 
 create_per_service_target_build_for_sample(keyvault sample1-hello-world)
 
-target_link_libraries(sample1-hello-world PRIVATE azure-security-keyvault-keys azure-identity get-env-helper)
+target_link_libraries(sample1-hello-world PRIVATE azure-security-keyvault-keys azure-security-attestation azure-identity get-env-helper)
```
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-keys/test/samples/sample1-hello-world/CMakeLists.txt:17]: Why would a keyvault hello world sample link `azure-security-attestation`?

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp [hunk 11]
```diff
@@ -292,6 +299,39 @@ namespace Azure { namespace Security { namespace KeyVault { namespace Keys {
...
+   *
+   */
+  class KeyEncryptionAlgorithm final
+      : public Azure::Core::_internal::ExtendableEnumeration<KeyEncryptionAlgorithm> {
```
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp:307]: @LarryOsterman, since you designed this type, is public inheritance from an  `_internal` type like this intended and acceptable?

File: eng/pipelines/templates/stages/platform-matrix-live.json (hunk 1 above)
Comment (gearama) [reply eng/pipelines/templates/stages/platform-matrix-live.json:131]: done

File: sdk/keyvault/azure-security-keyvault-certificates/CHANGELOG.md (hunk 2 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-certificates/CHANGELOG.md:14]: done

File: sdk/keyvault/azure-security-keyvault-certificates/inc/azure/keyvault/certificates/certificate_client_options.hpp (hunk 3 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-certificates/inc/azure/keyvault/certificates/certificate_client_options.hpp:31]: Yes, ask Jeff, we want to make it obvilus for consumers that while you can use different service version we do not guarantee the result, we only guarantee with the latest version :)

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp (hunk 5 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp:125]: done , fixed all

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_options.hpp (hunk 8 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_options.hpp:46]: Jeff wants this, it makes sense to have it this way. Having this struct and the versions defined implies we are testing and guaranteeing that thhe SDK works with other versions than 7.3 which is false, we don't even run the unit tests with previous versions. WE DO NOT GUARANTEE THAT OTHER VERSIONS THAN THE CURRENT ONE WORK.

File: sdk/keyvault/azure-security-keyvault-keys/test/samples/sample1-hello-world/CMakeLists.txt (hunk 10 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/test/samples/sample1-hello-world/CMakeLists.txt:17]: good catch , i was using it for some prototype .

File: sdk/keyvault/azure-security-keyvault-keys/vcpkg/Config.cmake.in [hunk 12]
```diff
@@ -4,7 +4,7 @@
 @PACKAGE_INIT@
 
 include(CMakeFindDependencyMacro)
-find_dependency(azure-core-cpp "1.3.1")
+find_dependency(azure-core-cpp "1.5.0")
```
[the remark this answers is not in the document]
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/vcpkg/Config.cmake.in:7]: yes, for rotation key

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp (hunk 11 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp:307]: he made me use it , so it better be ok, otherwise i'll get upset as code copmplete has passed.

File: sdk/keyvault/azure-security-keyvault-keys/test/ut/test_consts.hpp [hunk 13]
```diff
@@ -0,0 +1,235 @@
+#include <azure/core/internal/environment.hpp>
+#include <string>
+
+// cspell:disable
+namespace Azure { namespace Security { namespace KeyVault { namespace Keys { namespace Test {
+
+  const std::string AttestationServiceUrl()
```
[the remark this answers is not in the document]
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/test/ut/test_consts.hpp:7]: no

**Comment by gearama** [comment]
@ahsonkhan  your comments have been addressed , please review and if it 's ok approve so then i can merge and run the live tasts over the weekend

**Comment by ahsonkhan** [comment]
Thanks @gearama. Are we targeting this change for the July release?

I would like to discuss the ServiceVersion change some more, to understand the goal better. Let's find some time today for it.

cc @JeffreyRichter

File: sdk/keyvault/azure-security-keyvault-certificates/CHANGELOG.md [hunk 14]
```diff
@@ -4,11 +4,11 @@
...
 ### Features Added
 
-### Breaking Changes
+- Keyvault 7.3 support added for Certificates.
 
-### Bugs Fixed
+### Breaking Changes
 
-### Other Changes
+- Removed ServiceVersion type, replaced with Version field in the CertificateClientOptions type;
```
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-certificates/CHANGELOG.md:11]: Here and elsewhere
```suggestion
- Removed ServiceVersion type, replaced with Version field in the CertificateClientOptions type.
```

File: sdk/keyvault/azure-security-keyvault-certificates/inc/azure/keyvault/certificates/certificate_client_options.hpp [hunk 15]
```diff
@@ -18,59 +18,23 @@
...
     /**
      * @brief Construct a new Certificate Client Options object.
      *
-     * @param version Optional version for the client.
      */
-    CertificateClientOptions(ServiceVersion version = ServiceVersion::V7_2)
-        : Azure::Core::_internal::ClientOptions(), Version(version)
-    {
-    }
+    CertificateClientOptions() : Azure::Core::_internal::ClientOptions() { Version = "7.3"; }
```
[the remark this answers is not in the document]
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-certificates/inc/azure/keyvault/certificates/certificate_client_options.hpp:37]: Archboard review, Jeff asked fro it, so we make it clear to customers that the service version that is guaranteed is 7.3 , having a service version class and version defined implies that we have tested and guarantee that all versions specified in code work. which could not be further from the truth. We test and guarantee only one version

File: sdk/keyvault/azure-security-keyvault-certificates/test/ut/certificate_client_test.cpp [hunk 16]
```diff
@@ -888,3 +888,22 @@ TEST_F(KeyVaultCertificateClientTest, DISABLED_MergeCertificate)
...
+
+TEST_F(KeyVaultCertificateClientTest, ServiceVersion)
+{
+  auto credential
+      = std::make_shared<Azure::Identity::ClientSecretCredential>("tenantID", "AppId", "SecretId");
+  {
+    // 7.3
+    EXPECT_NO_THROW(auto options = CertificateClientOptions(); CertificateClient certificateClient(
+                        "http://account.vault.azure.net", credential, options);
+                    EXPECT_EQ(options.Version, "7.3"););
```
Comment (LarryOsterman) [inline sdk/keyvault/azure-security-keyvault-certificates/test/ut/certificate_client_test.cpp:900]: Line 900 seems to be confirming that the compiler correctly generated code. 

Do you have reason to believe that the compiler will incorrectly process the initializer for Version?

File: sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp [hunk 17]
```diff
@@ -211,5 +231,88 @@ TEST_F(KeyVaultKeyClient, CreateRsaHsmKey)
...
+    EXPECT_FALSE(keyResponse.Value.Properties.ReleasePolicy.HasValue());
+    EXPECT_TRUE(keyVaultKey.Properties.Enabled.Value());
+  }
+}
+
+std::string BinaryToHexString(std::vector<uint8_t> const& src)
+{
+  static constexpr char hexMap[]
+      = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};
+  std::string output(static_cast<size_t>(src.size()) * 2, ' ');
```
Comment (LarryOsterman) [inline sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp:243]: Why initialize with spaces? Why not zero initialize (`\0`)?

File: sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp [hunk 18]
```diff
@@ -211,5 +231,88 @@ TEST_F(KeyVaultKeyClient, CreateRsaHsmKey)
...
+{
+  auto const keyName = GetTestName();
+  auto const& client = GetClientForTest(keyName);
+
+  Azure::Security::KeyVault::Keys::CreateKeyOptions options;
+  options.KeyOperations.push_back(Azure::Security::KeyVault::Keys::KeyOperation::Sign);
+  options.KeyOperations.push_back(Azure::Security::KeyVault::Keys::KeyOperation::Verify);
+  options.ReleasePolicy = KeyReleasePolicy();
+  options.ReleasePolicy.Value().Immutable = false;
+  std::string dataStr = "{"
```
Comment (LarryOsterman) [inline sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp:265]: Use string literal here instead of concatenating the string (it makes it somewhat easier to read in printf statements and the debugger)?

**Comment by JeffreyRichter** [comment]
I requested the removal of ServiceVersion type and the hard-coding of the version string - I think we should do this in all our client libraries. Here's why: When we build a client library, we only test it with the latest service version and therefore, we can only support customers using this specific version. Different versions (newer or older) may be breaking and it is highly unlikely that our client library will work with a different service version. In which case, customers using this feature will cause our support burden to be very high.

By formalizing the service version concept with a specific data type and with some constant service version values, we are implying that this is a fully supported feature, and that the client library is **guaranteed** to work with older service versions. But we never test this, and we can't make this guarantee. By removing the ServiceVersion type and just having a simple string, we are not formalizing this concept, we are just **implying** to a customer that they could **_try_** a different string version and the client library may or may not work for them - our documentation for the service version string should even indicate that changing this value is not guaranteed to work. This way, customers understand that we have not tested this and do not support it. But, if it works for them and the specific operations and specific service versions they are trying, then great! We're happy to have made this possible for them.

File: sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp [hunk 19]
```diff
@@ -1,16 +1,28 @@
...
+#include <azure/core/internal/json/json.hpp>
 #include <azure/keyvault/keys.hpp>
 #include <private/key_constants.hpp>
-
 #include <string>
 
+using namespace Azure::Core::_internal;
+using namespace Azure::Security::KeyVault::Keys;
 using namespace Azure::Security::KeyVault::Keys::Test;
+using namespace Azure::Security::Attestation;
```
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp:22]: Why are we including attestation headers and namespaces in keyvault live tests?

We shouldn't have one SDK (source or test) take a dependency on another. It seems odd to me, but maybe I am missing something.

File: sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp [hunk 20]
```diff
@@ -1,16 +1,28 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // SPDX-License-Identifier: MIT
 
-#include "gtest/gtest.h"
-
+#include "../../azure-security-attestation/src/private/crypto/inc/crypto.hpp"
```
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp:4]: Why is this needed? At a glance, I don't see anything in that header we need in this file.

Each SDK should be self-contained and individually buildable/testable, other than a dependency on core (and identity). Attestation and Keyvault are fairly independent.

**Review by ahsonkhan — APPROVED** [review]
Beyond the service version change discussion, this PR looks good to me overall.

Given we are shipping a beta release, I recommend we keep the `ServiceVersion` class for now, to remain consistent with previously shipped client SDKs (and previous version of KeyVault), and revisit this more broadly. Because it is reversible, either way, if we feel strongly to keep this PR as is, in the interest of time, that works too :)

Though I understand your concern that there might be an implied support to the customer or possible mismatch of expectations, @JeffreyRichter, that can be addressed with similar documentation.

I'd like to discuss the trade-offs a bit, and if we really want to do so, come up with a strategy to update all SDKs going forward. And take our time with such a change, along with update the guidelines as necessary - https://azure.github.io/azure-sdk/cpp_introduction.html#service-versions.

**Comment by JeffreyRichter** [comment]
First, swaggers frequently have a mistake in them or lack of clarity that improving the swagger introduces a break which we have to allow because the change causes the swagger to more accurately reflect the true service behavior.

Second, Azure's policy DOES allow services to make breaking changes (we need to do this to be agile and competitive). The policy requires that the service team announce to customers that the latest version is breaking, and the old versions must work as-is for at least 3 more years. So, under this policy, we cannot guarantee that any client lib is backward compatible. 

Third, services occasionally break due to legal, security, or compliance issues. In these cases, the customer notice is just 1 year and sometimes less depending on the severity of the issue. These are rare but they do happen. 

In short: Service will break customers from time to time and we need to be cognizant of this.

**Comment by JeffreyRichter** [comment]
I'm not sure where the C++ guidance on ServiceVersion came from but we can (and should) change that (in my opinion).

**Comment by ahsonkhan** [comment]
> First, swaggers frequently have a mistake in them or lack of clarity that improving the swagger introduces a break which we have to allow because the change causes the swagger to more accurately reflect the true service behavior.
> 
> Second, Azure's policy DOES allow services to make breaking changes (we need to do this to be agile and competitive). The policy requires that the service team announce to customers that the latest version is breaking, and the old versions must work as-is for at least 3 more years. So, under this policy, we cannot guarantee that any client lib is backward compatible.
> 
> Third, services occasionally break due to legal, security, or compliance issues. In these cases, the customer notice is just 1 year and sometimes less depending on the severity of the issue. These are rare but they do happen.
> 
> In short: Service will break customers from time to time and we need to be cognizant of this.

Regarding 1, a swagger update to truly reflect the service behavior wouldn't really be a breaking change, so we wouldn't have to worry about that. As far as I understood things, we don't want to design our client SDK and versioning to try to paper over swagger or service-side breaking changes, anyway.

Regarding 2 & 3, since our SDK follows semver, if the service was to actually make breaking changes, we'd follow suite and rev the major version.

File: sdk/keyvault/azure-security-keyvault-certificates/test/ut/certificate_client_test.cpp (hunk 16 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-certificates/test/ut/certificate_client_test.cpp:900]: JUST TO MAKE SURE,

File: sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp (hunk 18 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp:265]: I don't see how that helps https://en.cppreference.com/w/cpp/language/string_literal 
it's exactly the same as what i did, it would still concatenate strings, and i prefer to have a formatted json vs one long interminable line, that would then be cut into pieces by the formatter, thus again concatenating strings.

File: sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp (hunk 19 above)
Comment (LarryOsterman) [reply sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp:22]: IIRC, KV's Secure Key Release requires using the attestation service. If we want a live test for SKR, you need to use the attestation service.

**Comment by LarryOsterman** [comment]
> I requested the removal of ServiceVersion type and the hard-coding of the version string - I think we should do this in all our client libraries. Here's why: When we build a client library, we only test it with the latest service version and therefore, we can only support customers using this specific version. Different versions (newer or older) may be breaking and it is highly unlikely that our client library will work with a different service version. In which case, customers using this feature will cause our support burden to be very high.
> 
> By formalizing the service version concept with a specific data type and with some constant service version values, we are implying that this is a fully supported feature, and that the client library is **guaranteed** to work with older service versions. But we never test this, and we can't make this guarantee. By removing the ServiceVersion type and just having a simple string, we are not formalizing this concept, we are just **implying** to a customer that they could **_try_** a different string version and the client library may or may not work for them - our documentation for the service version string should even indicate that changing this value is not guaranteed to work. This way, customers understand that we have not tested this and do not support it. But, if it works for them and the specific operations and specific service versions they are trying, then great! We're happy to have made this possible for them.

That makes sense. However given those restrictions, why do we have a service version construct in the public API surface at all? And should the service throw if it encounters a service version which *isn't* supported? That way if a client specifies version 4 of the API but the SDK doesn't support version 4, it throws an exception to let the client know that it cannot be supported.

**Comment by JeffreyRichter** [comment]
The reason to even offer the ServiceVersion field is that it provides an escape hatch for customers. C++ and many other (but not all) languages allow only one version of a client library to be loaded in a process and it's possible that the customer may want the old service behavior for some of their code and the new service behavior for other parts of their code. This "feature" allows the customer to TRY doing this. Without this feature, the customer MUST modify all their code to expect/use the new service behavior. If this fails, then the customer is forced to update all their code. 

Another example where the ServiceVersion field can be useful is when not all clouds support the latest version of a service. This is particularly true for Azure Stack where the customers "own" upgrading **_the_** Azure services on the customer's timeframe. Unlike public clouds where MS owns updating the services. For these scenarios, the customer HAS to choose the api-version that is installed/available to them.

Another (rare) example (related to the example above) is something that the AzCopy utility makes BIG usage of: AzCopy allows customers to copy blogs from Azure Stack to a public cloud and vice versa. If this tool were written in C++, it can only load 1 version of the C++ Blob client library, but it must send download requests to Azure Stack using one version and upload requests to the public cloud using another version - the ServiceVersion field makes this possible. And, since Storage remains largely backward compatible and doesn't require newer esoteric features, this actually works very well in practice.

**Comment by LarryOsterman** [comment]
> The reason to even offer the ServiceVersion field is that it provides an escape hatch for customers. C++ and many other (but not all) languages allow only one version of a client library to be loaded in a process and it's possible that the customer may want the old service behavior for some of their code and the new service behavior for other parts of their code. This "feature" allows the customer to TRY doing this. Without this feature, the customer MUST modify all their code to expect/use the new service behavior. If this fails, then the customer is forced to update all their code.
> 
> Another example where the ServiceVersion field can be useful is when not all clouds support the latest version of a service. This is particularly true for Azure Stack where the customers "own" upgrading **_the_** Azure services on the customer's timeframe. Unlike public clouds where MS owns updating the services. For these scenarios, the customer HAS to choose the api-version that is installed/available to them.
> 
> Another (rare) example (related to the example above) is something that the AzCopy utility makes BIG usage of: AzCopy allows customers to copy blogs from Azure Stack to a public cloud and vice versa. If this tool were written in C++, it can only load 1 version of the C++ Blob client library, but it must send download requests to Azure Stack using one version and upload requests to the public cloud using another version - the ServiceVersion field makes this possible. And, since Storage remains largely backward compatible and doesn't require newer esoteric features, this actually works very well in practice.

We should probably talk about that tomorrow AM (@ahsonkhan has made a similar request). IMHO, the only time ServiceVersion makes sense is if the client supports multiple versions. And then the client needs to reject requests to create the client with API versions that aren't supported by the client (since there is a presumption that there are breaking changes to the service API associated with each service API version). If the client doesn't understand the ServiceVersion provided by the customer, then the client should reject the ServiceVersion. And yes, that means that it's possible that if we update the client to remove support for a particular API version, it means we may force customers to change their code.

If the client is expected to support multiple versions, then IMHO it is incumbent on the client to ensure that it is tested against each of the versions it supports. There should be no situations where the client supports multiple versions but is tested against only one of them.

File: sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp (hunk 17 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp:243]: done

File: sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp (hunk 18 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp:265]: fiiiine , done :) you got me at debug

File: sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp (hunk 20 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/test/ut/key_client_create_test_live.cpp:4]: done
